### PR TITLE
Use more secure error for invalid pagination ranges

### PR DESCRIPTION
**Pull request by linc**
* Disallow limit and offset clauses too large for MySQL.
* Make better use of the Offset method rather than allowing it to be touched directly elsewhere.

### CODE REVIEW

File: library/database/class.sqldriver.php [hunk 1]
```diff
@@ -1283,10 +1283,16 @@ public function Like($Field, $Match = '', $Side = 'both', $Op = 'like') {
     * @return Gdn_SQLDriver $this
     */
    public function Limit($Limit, $Offset = FALSE) {
+      // SQL chokes on ints over 2^31
+      if ($Limit > 2147483648) {
```
Comment (tburry) [inline library/database/class.sqldriver.php]: Put this in a class constant please.

Comment (linc) [inline library/database/class.sqldriver.php]: Done, updated

File: library/database/class.sqldriver.php [hunk 2]
```diff
@@ -1263,10 +1266,16 @@ public function Like($Field, $Match = '', $Side = 'both', $Op = 'like') {
     * @return Gdn_SQLDriver $this
     */
    public function Limit($Limit, $Offset = FALSE) {
+      // SQL chokes on ints over 2^31
+      if ($Limit > MAX_SIGNED_INT) {
+         throw new Exception(T('Invalid limit defined.'), 400);
```
Comment (tburry) [inline library/database/class.sqldriver.php]: 'Invalid limit.' instead please. The "defined" part seems too bleep bloop.

File: library/database/class.sqldriver.php [hunk 3]
```diff
@@ -1367,6 +1376,11 @@ public function NoReset($NoReset = TRUE, $OneTime = TRUE) {
     * @return Gdn_SQLDriver $this
     */
    public function Offset($Offset) {
+      // SQL chokes on ints over 2^31
+      if ($Offset > MAX_SIGNED_INT) {
+         throw new Exception(T('Invalid offset defined.'), 400);
```
Comment (tburry) [inline library/database/class.sqldriver.php]: 'Invalid offset.'

Comment (kaecyra) [inline library/database/class.sqldriver.php]: image

**Comment by linc** [comment]
Would appreciate it if, in general, we could catch more things on the first round of feedback rather than nickel-and-diming as we go, especially things like this that didn't change. Revisiting these is significant overhead.

**Pull request merged**